src/lib/metrics_extractor/base_metrics_extractor.py

```diff
@@ -1,36 +1,49 @@
 from abc import ABC, abstractmethod
 from datetime import datetime
 
-from lib.aws.timestream_manager import TimestreamTableWriter
+from lib.aws.timestream_manager import TimestreamTableWriter, TimeStreamQueryRunner
 
 class BaseMetricsExtractor(ABC):
     """
     Base Class which provides unified functionality for extracting metrics
 
     Attributes:
         aws_service_client: Boto3 AWS Service client
-        entity_name (str): Name of the entity (e.g. Glue job) we are extracting metrics for        
+        resource_name (str): Name of the resource (e.g. Glue job, Lambda Function etc.) we are extracting metrics for        
         monitored_environment_name (str): Name of the monitored environment
         timestream_db_name (str): Name of the Timestream DB (where metrics are written to)
         timestream_metrics_table_name (str): Name of the Timestream table (where metrics are written to)                
     """ 
 
+    RESOURCE_NAME_COLUMN_NAME = 'resource_name'
+    EXECUTION_MEASURE_NAME = 'execution'
 
     def __init__(
-        self, aws_service_client, entity_name, monitored_environment_name, timestream_db_name, timestream_metrics_table_name
+        self, aws_service_client, resource_name, monitored_environment_name, timestream_db_name, timestream_metrics_table_name
     ):
         self.aws_service_client = aws_service_client
-        self.entity_name = entity_name
+        self.resource_name = resource_name
         self.monitored_environment_name = monitored_environment_name
         self.timestream_db_name = timestream_db_name
         self.timestream_metrics_table_name = timestream_metrics_table_name
 
-    @abstractmethod
     def get_last_update_time(self, timestream_query_client) -> datetime:
         """
         Get time of this entity's data latest update (we append data since that time only)
         """
-        pass
+        queryRunner = TimeStreamQueryRunner(
+            timestream_query_client=timestream_query_client
+        )
+
+        # check if table is empty
+        query = f'SELECT count(*) FROM "{self.timestream_db_name}"."{self.timestream_metrics_table_name}"'
+        count_rec = int(queryRunner.execute_scalar_query(query=query))
+        if count_rec == 0:
+            return None
+
+        query = f'SELECT max(time) FROM "{self.timestream_db_name}"."{self.timestream_metrics_table_name}" WHERE {self.RESOURCE_NAME_COLUMN_NAME} = \'{self.resource_name}\''
+        last_date = queryRunner.execute_scalar_query_date_field(query=query)
+        return last_date
```

Comment:
A bit confused why we need this method also in BaseMetricsExtractor, seems only the one from metrics_extractor_utils is being used :)

Reply:
Actually, both are used in lambda_extract_metrics. 
And it was the idea - if we don't get last_update_times from orchestrator lambda (e.g. we ran extract lambda manually for debug, it takes last_update_time from timestream table directly, using this method):

                # 2. Get time of this entity's data latest update (we append data since that time only)
                since_time = get_last_update_time(
                    last_update_times, resource_type, resource_name
                )
                ...

                if since_time is None:
                    # if last_update_time was not given or missing - query for specific resource directly from table
...
                    since_time = metrics_extractor.get_last_update_time(
                        timestream_query_client=timestream_query_client
                    )

Reply:
I see now, thanks! Yes, it is quite useful then, let's leave as it is